client-worker/src/lib.rs

```diff
@@ -9,137 +9,217 @@ connected_.  Outside of their type, which is checked at call time,
+    async fn apply_client_command<Fut, T>(
+        &self,
+        chain_id: ChainId,
+        mut f: impl FnMut(&mut ChainClient) -> Fut,
+    ) -> Result<T, JsError>
+    where
+        Fut: Future<Output = Result<ClientOutcome<T>, JsError>>,
+    {
+        loop {
+            let mut chain_client = self.client_context.lock().await.make_chain_client(chain_id)?;
+            let mut stream = chain_client.subscribe().await?;
+            let result = f(&mut chain_client).await;
+            self.client_context.lock().await.update_wallet(&chain_client).await?;
+            let timeout = match result? {
+                ClientOutcome::Committed(t) => return Ok(t),
+                ClientOutcome::WaitForTimeout(timeout) => timeout,
+            };
+            drop(chain_client);
+            linera_client::util::wait_for_next_round(&mut stream, timeout).await;
         }
     }
 
-    Ok(())
-}
+    #[wasm_bindgen(constructor)]
+    pub async fn new(wallet: JsWallet) -> Result<Client, JsError> {
+        let JsWallet(wallet) = wallet;
+        let mut storage = get_storage().await?;
+        wallet.genesis_config().initialize_storage(&mut storage).await?;
+        let client_context = Arc::new(AsyncMutex::new(ClientContext::new(storage.clone(), OPTIONS, wallet)));
+        ChainListener::new(ChainListenerConfig::default())
+            .run(client_context.clone(), storage)
+            .await;
+        log::info!("Linera Web client successfully initialized");
+        Ok(Self {
+            client_context,
+        })
+    }
 
-#[wasm_bindgen]
-pub async fn set_wallet(wallet: &str) -> Result<(), wasm_bindgen::JsError> {
-    linera_client::config::WalletState::create_from_local_storage("linera-wallet", serde_json::from_str(wallet)?)?;
-    Ok(())
+    #[wasm_bindgen]
+    pub fn frontend(&self) -> Frontend {
+        Frontend(self.clone())
+    }
 }
 
+// A serializer suitable for serializing responses to JavaScript to be
+// sent using `postMessage`.
+static RESPONSE_SERIALIZER: serde_wasm_bindgen::Serializer =
+    serde_wasm_bindgen::Serializer::new()
+        .serialize_large_number_types_as_bigints(true)
+        .serialize_maps_as_objects(true);
+
 #[wasm_bindgen]
-pub async fn dapp_query(n: u32) -> u32 {
-    n + 1
+impl Frontend {
+    #[wasm_bindgen]
+    pub async fn validator_version_info(&self) -> Result<JsValue, JsError> {
+        let mut client_context = self.0.client_context.lock().await;
+        let chain_id = client_context.wallet().default_chain().expect("No default chain");
+        let chain_client = self.0.client_context.lock().await.make_chain_client(chain_id)?;
+        chain_client.synchronize_from_validators().await?;
+        let result = chain_client.local_committee().await;
+        client_context.update_and_save_wallet(&chain_client).await?;
+        let committee = result?;
+        let node_provider = client_context.make_node_provider();
+
+        let mut validator_versions = HashMap::new();
```

Comment:
Since the HashMap gets serialized into the output, the output could vary from one run to another. So, maybe it is better to use `BTreeMap` here.

Reply:
It's being serialized into a JavaScript object, which anyway doesn't preserve property order — so I don't think it's useful to use a `BTreeMap` in this case.